0022-circuit-block.md

```diff
@@ -0,0 +1,359 @@
+The case-study use of this is for Qiskit Runtime primitives to provide more flexibility and transparency in Pauli twirling and error mitigation.
+
+Qiskit does not currently have the concept of a "grouping" of instructions, other than as a user-defined custom gate.
+However, the ability to group certain instructions to appear as a "block" in a larger context is important for many compilation tasks:
+
+* a block of several gates might have a known unitary action, and can be treated as atomic for commutation purposes, even if a candidate gate doesn't commute with each block element individually.
+* noise-learning techniques might be applied to simultaneous logical blocks of operations, rather than individual 2q hardware gates.
+* a user might wish to Pauli twirl around a block of instructions, including idling wires, rather than a single one.
+* a block of instructions might be assigned some duration as part of a larger circuit, and internally it is scheduled with stretchable durations to implement optimal dynamic decoupling spacing without increasing the complexity of the outer scheduling problem.
+
+Several downstream projects of Qiskit have made ad-hoc blocks by using labelled `Barrier` instructions, but this optimisation barrier has further effects than simple grouping, and the single-sided nature of it makes it very hard to work with as an actual "grouping" construct.
+`Barrier`s also are hard for users to interpret when inspecting circuits after compilation.
+
+The RFC proposes to add a new `Box` instruction to Qiskit, whose semantics will be very similar to the `box`
+[concept from OpenQASM 3](https://openqasm.com/language/scope.html#block-scope).
+This `Box` should be able to pass through the Qiskit compiler (under certain limitations), be QPY serialisable, and be able to be transmitted up and down execution stacks.
+We will examine how this `Box` might also permit downstream "annotations", outside of the control of the Qiskit SDK compiler, and how the compiler will still be able to reason about such annotations.
+
+The nearest-term use cases for this are:
+
+* have a grouping construction to simplify stretch-based scheduling in Qiskit SDK
+* allow advanced users to specify the noise-learning and twirling blocks IBM Primitives error mitigation should use, and the per-block strategy
+* allow regular users to see what blocks the IBM Primitives divided their circuit into for noise learning and twirling
+
+
+## `Box`: a one-off grouping of instructions
+
+As in OpenQASM 3, a `Box` will be:
+
+* a group of non-overlapping (but potentially nested) circuit instructions,
+* which introduces a new scope for declared variables,
+* which can have internal scheduling within itself, but is scheduled atomically within the circuit (i.e. all wires touched within the box are delayed to match the critical path length),
+* which forbids optimisations from crossing from outside to inside the box or vice-versa, but permits optimisations across the entire box,
```

Comment:
The regular and power workflows below assume that optimization happens before the assignment of boxes.

Reply:
Although we want the boxes for twirling and mitigation, this RFC defines them more generally, as a core generic construct in qiskit. So the workflows below are specific to the use-case of twirling, but the construct itself needs to have this sort of generic interaction with the transpiler well-defined.